### PR TITLE
Add support for RFXCOM undecoded messages

### PR DESCRIPTION
**Pull request by Jamstah**
Support for undecoded message is already in 1.8.3, this is a port of that functionality to the version 2 binding.

### CODE REVIEW

**Comment by kaikreuzer** [comment]
Lgtm, thanks.
@paulianttila Could you please also review and comment?

**Comment by Jamstah** [comment]
Have just addressed merge conflicts and rebased this pull request so its ready for master.

**Comment by Jamstah** [comment]
Closed during rebase somehow, probably because it went through a stage of having 0 additional commits.

**Comment by Jamstah** [comment]
Just done a second merge/rebase, this is a few months old now. @paulianttila / @kaikreuzer, can this get merged?

**Comment by Jamstah** [comment]
@martinvw would you mind reviewing this one for me?

File: addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/messages/RFXComUndecodedRFMessage.java [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+        ARC(0x01),
+        ATI(0x02),
+        HIDEKI_UPM(0x03),
+        LACROSSE_VIKING(0x04),
+        AD(0x05),
+        MERTIK(0x06),
+        OREGON1(0x07),
+        OREGON2(0x08),
+        OREGON3(0x09),
+        PROGUARD(0x0a),
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/messages/RFXComUndecodedRFMessage.java]: _Upper or lowercase, I would prefer uppercase_

File: addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/messages/RFXComUndecodedRFMessage.java [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+
+        SubType(byte subType) {
+            this.subType = subType;
+        }
+
+        public byte toByte() {
+            return (byte) subType;
+        }
+
+        public static SubType fromByte(int input) {
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/messages/RFXComUndecodedRFMessage.java]: Nice way to handle this ;-)

File: addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/messages/RFXComUndecodedRFMessage.java [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+        super.encodeMessage(message);
+
+        subType = SubType.fromByte(super.subType);
+        rawPayload = Arrays.copyOfRange(rawMessage, 4, rawMessage.length);
+    }
+
+    @Override
+    public byte[] decodeMessage() {
+
+        final int rawPayloadLen = Math.min(rawPayload.length, 33);
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/messages/RFXComUndecodedRFMessage.java]: Do you want to issue some warning when you discard some piece of payload?

File: addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/messages/RFXComUndecodedRFMessage.java [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+        byte[] data = new byte[4 + rawPayloadLen];
+
+        data[0] = (byte) (data.length - 1);
+        data[1] = RFXComBaseMessage.PacketType.UNDECODED_RF_MESSAGE.toByte();
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/messages/RFXComUndecodedRFMessage.java]: _I prefer static imports_

File: addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/messages/RFXComUndecodedRFMessage.java [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+    }
+
+    @Override
+    public void setDeviceId(String deviceId) throws RFXComException {
+        throw new RFXComException("Not supported");
+    }
+
+    @Override
+    public void convertFromState(RFXComValueSelector valueSelector, Type type) throws RFXComException {
+        throw new RFXComException("Not supported");
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/messages/RFXComUndecodedRFMessage.java]: could be possible to send an item as well using this protocol?

File: addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/messages/RFXComUndecodedRFMessage.java [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+
+        for (SubType s : SubType.values()) {
+            if (s.toString().equals(subType)) {
+                return s;
+            }
+        }
+
+        // try to find sub type by number
+        try {
+            return SubType.values()[Integer.parseInt(subType)];
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/messages/RFXComUndecodedRFMessage.java]: O oh you use the array index here instead of the byte, for now the array indices match the bytes but no long term guarantees, use the static method please.

Comment (martinvw) [inline addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/messages/RFXComUndecodedRFMessage.java]: Because the exception you are catching there is an ArrayIndexOutOfBoundsExecption which you should not catch (or throw)

File: addons/binding/org.openhab.binding.rfxcom/ESH-INF/thing/channels.xml [hunk 7]
```diff
@@ -5,6 +5,18 @@
 
 	<!-- Channel definitions -->
 
+    <channel-type id="rawmessage">
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.rfxcom/ESH-INF/thing/channels.xml]: It looks like this formatted a little bit different then the rest of the file. Did you format it with the ESH profile?

File: addons/binding/org.openhab.binding.rfxcom/ESH-INF/thing/undecoded.xml [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+			        <option value="OREGON1">Oregon Scientific 1</option>
+			        <option value="OREGON2">Oregon Scientific 2</option>
+			        <option value="OREGON3">Oregon Scientific 3</option>
+			        <option value="PROGUARD">ProGuard</option>
+			        <option value="VISONIC">Visonic</option>
+			        <option value="NEC">NEC</option>
+			        <option value="FS20">FS20</option>
+			        <option value="RESERVED">Reserved</option>
+			        <option value="BLINDS">Blinds</option>
+			        <option value="RUBICSON">RUBiCSON</option>
```
Comment (Jamstah) [inline addons/binding/org.openhab.binding.rfxcom/ESH-INF/thing/undecoded.xml]: Copy paste from v1 binding, but a google search suggests normal case is more appropriate.

File: addons/binding/org.openhab.binding.rfxcom/README.md [hunk 9]
```diff
@@ -104,6 +105,8 @@ This binding currently supports following channels:
 | motion | Switch | Motion detection sensor state. |
 | rainrate | Number | Rain fall rate in millimeters per hour. |
 | raintotal | Number | Total rain in millimeters. |
+| rawmessage | String | Hexadecimal string of the raw RF message. |
```
Comment (Jamstah) [inline addons/binding/org.openhab.binding.rfxcom/README.md]: Yes, you have to turn them on in the RFXCOM config, but that's the same for any device, it has to be enabled in the config. I don't think it needs expicitly calling out, although it might not be a bad idea to make rawmessage and rawpayload work for all messages, regardless of type.

File: addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/messages/RFXComUndecodedRFMessage.java (hunk 2 above)
Comment (Jamstah) [inline addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/messages/RFXComUndecodedRFMessage.java]: Haha, I nabbed that from your pending pull request, because I'm hoping it goes in eventually too.

File: addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/messages/RFXComUndecodedRFMessage.java (hunk 3 above)
Comment (Jamstah) [inline addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/messages/RFXComUndecodedRFMessage.java]: Not sure what you mean by discarding a piece of payload? That is just making sure the message is at least the minimum length, not cutting off a message that is too long.

File: addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/messages/RFXComUndecodedRFMessage.java (hunk 4 above)
Comment (Jamstah) [inline addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/messages/RFXComUndecodedRFMessage.java]: Good call, especially as that one is ugly and doesn't even need the RFXComBaseMessage as its extended. Have changed them all.

File: addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/messages/RFXComUndecodedRFMessage.java (hunk 5 above)
Comment (Jamstah) [inline addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/messages/RFXComUndecodedRFMessage.java]: Potentially, but I don't have anything to test sending with.

File: addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/RFXComBindingConstants.java [hunk 10]
```diff
@@ -64,6 +64,8 @@
             BRIDGE_RFXTRX315, BRIDGE_RFXREC443);
 
     // List of all Channel ids
+    public final static String CHANNEL_RAW_MESSAGE = "rawMessage";
+    public final static String CHANNEL_RAW_PAYLOAD = "rawPayload";
```
Comment (Jamstah) [inline addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/RFXComBindingConstants.java]: I think it does, but could be missing something. Can you be more specific?

File: addons/binding/org.openhab.binding.rfxcom/ESH-INF/thing/channels.xml (hunk 7 above)
Comment (Jamstah) [inline addons/binding/org.openhab.binding.rfxcom/ESH-INF/thing/channels.xml]: I had spaces instead of tabs. I guess the xml doesn't autoformat, fixed.

This is something that probably needs adding to the automatic eclipse config. The editor is set up with spaces for tabs, so I can't add a tab manually, but the formatter changes them on request.

**Comment by Jamstah** [comment]
The binding in general is for supporting "undecoded" items, which is basically anything that the firmware for the rfxcom doesn't understand. For me, I'm using it to pick up messages from super-cheap unbranded contact sensors that are really designed to be used with home security systems.

File: addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/RFXComBindingConstants.java (hunk 10 above)
Comment (martinvw) [inline addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/RFXComBindingConstants.java]: I refer to capital M/P there are some (closed) pull requests for the rain-message because there was something wrong with the caps there.

Here you have  _rawMessage_ in the xml you have _rawmessage_

File: addons/binding/org.openhab.binding.rfxcom/README.md (hunk 9 above)
Comment (martinvw) [inline addons/binding/org.openhab.binding.rfxcom/README.md]: What do you mean with 

> although it might not be a bad idea to make rawmessage and rawpayload work for all messages, regardless of type.

File: addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/messages/RFXComUndecodedRFMessage.java (hunk 3 above)
Comment (martinvw) [inline addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/messages/RFXComUndecodedRFMessage.java]: If the input is longer the 33 character you will not iterate over the content in the for-loop below and thus you will discard some input (when sending this message to the RFXCom, which I believe to be theoretically anyway)

File: addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/messages/RFXComUndecodedRFMessage.java (hunk 5 above)
Comment (martinvw) [inline addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/messages/RFXComUndecodedRFMessage.java]: Maybe you only enable this protocol and capturing something and then send the same string back to work with unsupported devices:

Globally it could look like this (untested could this work???):

String undecoded { rfxcom=">UNDECODED:UNDECODED.ARC:RawPayload" }

from a rule postCommand(undecoded, "HEXSTRINGHERE")

File: addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/messages/RFXComUndecodedRFMessage.java (hunk 6 above)
Comment (martinvw) [inline addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/messages/RFXComUndecodedRFMessage.java]: I see its OH2 specific code

Comment (martinvw) [inline addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/messages/RFXComUndecodedRFMessage.java]: Throwing the exception from the enum might to destructive for other code and we still have to do something if its not numeric. What do you think about the following fragment

```java
 public Object convertSubType(String subType) throws RFXComException {

        for (SubType s : SubType.values()) {
            if (s.toString().equals(subType)) {
                return s;
            }
        }

        // try to find sub type by number
        if (StringUtils.isNumeric(subType)){
            SubType fromByte = SubType.fromByte(Integer.parseInt(subType));
            if (fromByte != SubType.UNKNOWN) {
                return fromByte;
            }
        }
        
        throw new RFXComException("Unknown sub type " + subType);
    }

```

**Comment by martinvw** [comment]
My code with the remark for the convertSubType has just been merged, so you also add a test

File: addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/RFXComBindingConstants.java (hunk 10 above)
Comment (Jamstah) [inline addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/RFXComBindingConstants.java]: I think that gets mapped in undecoded.xml:
		<channels>
			<channel id="rawMessage" typeId="rawmessage" />
			<channel id="rawPayload" typeId="rawpayload" />
		</channels>

File: addons/binding/org.openhab.binding.rfxcom/README.md (hunk 9 above)
Comment (Jamstah) [inline addons/binding/org.openhab.binding.rfxcom/README.md]: I mean that all bindings could have a rawmessage and rawpayload channel if we wanted, so we could expose the raw bytes to rules if they want them. Would be another PR if it got done anyway.

File: addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/messages/RFXComUndecodedRFMessage.java (hunk 5 above)
Comment (Jamstah) [inline addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/messages/RFXComUndecodedRFMessage.java]: Yes, quite possibly. I suppose I could also do a test by capturing a message sent by a supported device and seeing if I could send it over the unsupported channel as raw bytes.

I won't do it for this PR, maybe as a future enhancement.

File: addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/messages/RFXComUndecodedRFMessage.java (hunk 6 above)
Comment (Jamstah) [inline addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/messages/RFXComUndecodedRFMessage.java]: I would rather put the exception in fromByte and see if the tests pass. At the moment, openhab2 is beta, so more risky refactoring should be done now and not left as technical debt for later. For this PR, I'd be tempted to leave as is, and then fix them all up the same way, whichever way is decided.

File: addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/messages/RFXComUndecodedRFMessage.java (hunk 5 above)
Comment (martinvw) [inline addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/messages/RFXComUndecodedRFMessage.java]: 👍

File: addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/messages/RFXComUndecodedRFMessage.java (hunk 3 above)
Comment (martinvw) [inline addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/messages/RFXComUndecodedRFMessage.java]: The RFXCom specifies a maximum payload of 33 so it's correct but this method 'provides' binary input by converting a rawPayload to a message. If the rawPayload is to big you silently (!) discard part of the rawPayload.

File: addons/binding/org.openhab.binding.rfxcom/README.md (hunk 9 above)
Comment (martinvw) [inline addons/binding/org.openhab.binding.rfxcom/README.md]: 👍

**Comment by martinvw** [comment]
Can you decide about limiting the amount of data to be converted, since this only relevant if we can use as output a code comment might suffice.

And can please add a test?

File: addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/messages/RFXComUndecodedRFMessage.java (hunk 3 above)
Comment (Jamstah) [inline addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/messages/RFXComUndecodedRFMessage.java]: OK great, if its in the spec, then we have a reason for it. I'll comment the code and produce a warning.

File: addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/messages/RFXComUndecodedRFMessage.java [hunk 11]
```diff
@@ -0,0 +1,201 @@
...
+import org.eclipse.smarthome.core.library.types.StringType;
+import org.eclipse.smarthome.core.types.State;
+import org.eclipse.smarthome.core.types.Type;
+import org.openhab.binding.rfxcom.RFXComValueSelector;
+import org.openhab.binding.rfxcom.internal.exceptions.RFXComException;
+
+/**
+ * RFXCOM data class for undecoded messages.
+ *
+ * @author James Hewitt-Thomas
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/messages/RFXComUndecodedRFMessage.java]: If you are backporting does it maybe also require the original author?

Comment (Jamstah) [inline addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/messages/RFXComUndecodedRFMessage.java]: Yes, I was thinking the same thing myself, will add.

**Comment by Jamstah** [comment]
Yes, I can, should, and will be adding a test.

**Comment by Jamstah** [comment]
Test added, cleaned up some warnings from the test package too, and Eclipse did some reformatting for me. Take a look, and I'll squash the commits when you're done.

File: addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/messages/RFXComUndecodedRFMessage.java (hunk 3 above)
Comment (martinvw) [inline addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/messages/RFXComUndecodedRFMessage.java]: 👍

**Comment by martinvw** [comment]
Great it looks good!

Please think along with me about the exceptions on the forum, then we can do it right on all places.

**Comment by Jamstah** [comment]
@kaikreuzer / @paulianttila. This is all squashed, rebased on Martin's earlier merge, and ready to go / for further review.

**Comment by Jamstah** [comment]
Thanks Kai. In preparation, I've done a rebase and retest to make sure the other committed changes didn't conflict. Its all good to go.

File: addons/binding/org.openhab.binding.rfxcom/ESH-INF/thing/undecoded.xml [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+
+		<channels>
+			<channel id="rawMessage" typeId="rawmessage" />
+			<channel id="rawPayload" typeId="rawpayload" />
+		</channels>
+
+		<config-description>
+			<parameter name="deviceId" type="text" required="true">
+				<label>Device Id</label>
+				<description>Sensor Id. Example 2983</description>
```
Comment (paulianttila) [inline addons/binding/org.openhab.binding.rfxcom/ESH-INF/thing/undecoded.xml]: Should description define that Device Id should be always UNDECODED?

Comment (Jamstah) [inline addons/binding/org.openhab.binding.rfxcom/ESH-INF/thing/undecoded.xml]: I don't see why not, have updated the XML.

**Comment by paulianttila** [comment]
Lgtm as well.

**Comment by kaikreuzer** [comment]
Perfect, thank you all!
